compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,534 @@
+                    if !tys.is_empty() {
+                        s.push('I');
+                        for ty in tys {
+                            s.push_str(&encode_ty(tcx, ty, dict, options));
+                        }
+                        s.push('E');
+                    }
+                }
+                s = format!("u{}{}", s.len(), &s);
+                compress(dict, (ty, TyQ::None), &mut s);
+            }
+            typeid.push_str(&s);
+        }
+
+        ty::Foreign(def_id) => {
+            // <length><name>, where <name> is <unscoped-name>
+            let mut s = String::new();
+            let name = format!("{}", tcx.item_name(*def_id));
+            s.push_str(&format!("{}", name.len()));
+            s.push_str(&name);
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        // Function types
+        ty::FnDef(def_id, ..) => {
+            // u2fnI<length><name>E as vendor extended type
+            let mut s = String::from("u2fnI");
+            s.push_str(&encode_ty_name(tcx, *def_id));
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Closure(_, substs) | ty::Generator(_, substs, _) => {
+            // u7closureI<return-type><element-type1..element-typeN>E as vendor extended type
+            let mut s = String::from("u7closureI");
+            let sig = substs.as_closure().sig();
+            // Push str without the initial 'F' but with the closing 'E'
+            s.push_str(&(encode_fnsig(tcx, &sig.skip_binder(), dict)[1..]));
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        // Pointer types
+        ty::Ref(_, ty0, mutbl) => {
+            // [U3mut]u3refI<element-type>E as vendor extended type qualifier and type
+            let mut s = String::new();
+            s.push_str("u3refI");
+            if options.contains(EncodeTyOptions::GENERALIZE_POINTERS) {
+                // Drop extended type qualifiers (i.e., make all references compatible)
+                s.push_str("vE");
+                compress(
+                    dict,
+                    (tcx.mk_imm_ref(tcx.lifetimes.re_static, tcx.mk_unit()), TyQ::None),
+                    &mut s,
+                );
+            } else {
+                s.push_str(&encode_ty(tcx, *ty0, dict, options));
+                s.push('E');
+                compress(dict, (ty, TyQ::None), &mut s);
+                if hir::Mutability::Mut == *mutbl {
+                    s = format!("{}{}", "U3mut", &s);
+                    compress(dict, (ty, TyQ::Mut), &mut s);
+                }
+            }
+            typeid.push_str(&s);
+        }
+
+        ty::RawPtr(tm) => {
+            // P[K]<element-type>
+            let mut s = String::new();
+            if options.contains(EncodeTyOptions::GENERALIZE_POINTERS) {
```

Comment:
> The `GENERALIZE_POINTERS` option is intended to offer behaviour compatible with `-fsanitize-cfi-icall-generalize-pointers`, right?
Yes.

> 
> In terms of implementation strategy, I think it should be a type level transformation, independent of an actual mangling process.

While generalize pointers was made specifically to solve a problem with Clang's use of Itanium C++ ABI mangling for CFI (see https://clang.llvm.org/docs/ControlFlowIntegrity.html#fsanitize-cfi-icall-generalize-pointers), I see the benefits of doing a type transformation instead, but this would still belong to and depend on the encoding (e.g., other encodings may drop type and extended type qualifiers, generalize all pointer types, etc.).

> Additionally, like in Clang it should apply only to the parameters with pointer types and pointer return type. In contrast to the current implementation which generalizes all pointer types, everywhere.

In the current implementation, the `GENERALIZE_POINTERS` option (which is not enabled by default and doesn't have a driver and frontend option yet) will be set for `typeid_for_fnabi` and `typeid_for_fnsig` only, when the driver and frontend option is added and the feature is enabled.